src/plugins/expressions/common/expression_functions/specs/moving_average.ts

```diff
@@ -0,0 +1,156 @@
+ * * If null or undefined value is encountered, skip the current row and do not change the window
+ * * For all values besides `null` and `undefined`, the value will be cast to a number before it's used in the
+ *   calculation of the current series even if this results in `NaN` (like in case of objects).
+ * * To determine separate series defined by the `by` columns, the values of these columns will be cast to strings
+ *   before comparison. If the values are objects, the return value of their `toString` method will be used for comparison.
+ */
+export const movingAverage: ExpressionFunctionMovingAverage = {
+  name: 'moving_average',
+  type: 'datatable',
+
+  inputTypes: ['datatable'],
+
+  help: i18n.translate('expressions.functions.movingAverage.help', {
+    defaultMessage: 'Calculates the moving average of a column in a data table',
+  }),
+
+  args: {
+    by: {
+      help: i18n.translate('expressions.functions.movingAverage.args.byHelpText', {
+        defaultMessage: 'Column to split the moving average calculation by',
+      }),
+      multi: true,
+      types: ['string'],
+      required: false,
+    },
+    inputColumnId: {
+      help: i18n.translate('expressions.functions.movingAverage.args.inputColumnIdHelpText', {
+        defaultMessage: 'Column to calculate the moving average of',
+      }),
+      types: ['string'],
+      required: true,
+    },
+    outputColumnId: {
+      help: i18n.translate('expressions.functions.movingAverage.args.outputColumnIdHelpText', {
+        defaultMessage: 'Column to store the resulting moving average in',
+      }),
+      types: ['string'],
+      required: true,
+    },
+    outputColumnName: {
+      help: i18n.translate('expressions.functions.movingAverage.args.outputColumnNameHelpText', {
+        defaultMessage: 'Name of the column to store the resulting moving average in',
+      }),
+      types: ['string'],
+      required: false,
+    },
+    window: {
+      help: i18n.translate('expressions.functions.movingAverage.args.windowHelpText', {
+        defaultMessage: 'The size of window to "slide" across the histogram.',
+      }),
+      types: ['number'],
+      required: false,
+      default: 5,
+    },
+  },
+
+  fn(input, { by, inputColumnId, outputColumnId, outputColumnName, window = 5 }) {
+    const resultColumns = buildResultColumns(
+      input,
+      outputColumnId,
+      inputColumnId,
+      outputColumnName
+    );
+
+    if (!resultColumns) {
+      return input;
+    }
+
+    const lastNValuesByBucket: Partial<Record<string, number[]>> = {};
+    return {
+      ...input,
+      columns: resultColumns,
+      rows: input.rows.map((row) => {
+        const newRow = { ...row };
+        const bucketIdentifier = getBucketIdentifier(row, by);
+        const lastNValues = lastNValuesByBucket[bucketIdentifier];
+        const currentValue = newRow[inputColumnId];
+        if (lastNValues != null && lastNValues.length && currentValue != null) {
```

Comment:
nit: I think the check `&& lastNValues.length` will never be false if `lastNValues != null` is true because the array is always initialized with at least one value and afterwards it's only values getting added.